locust_plugins/jmeter_listener.py

```diff
@@ -0,0 +1,250 @@
+        self.results_filename = (
+            "results_" + datetime.fromtimestamp(time()).strftime(self.results_timestamp_format) + ".csv"
+        )
+
+        # fields set by default in jmeter
+        self.csv_headers = [
+            "timeStamp",
+            "elapsed",
+            "label",
+            "responseCode",
+            "responseMessage",
+            "threadName",
+            "dataType",
+            "success",
+            "failureMessage",
+            "bytes",
+            "sentBytes",
+            "grpThreads",
+            "allThreads",
+            "Latency",
+            "IdleTime",
+            "Connect",
+        ]
+        self.results_file = self._create_results_log()
+        self.user_count = 0
+        self.user_name = ""
+        events.quitting.add_listener(self._write_final_log)
+        events.init.add_listener(self.on_locust_init)
+
+    def _set_user_name(self, name):
+        self.user_name = name
+
+    def _add_user(self):
+        self.user_count += 1
+
+    def on_locust_init(self, environment, **kwargs):
+        self.env = environment
+        user_classes = self.env.user_classes
+        self.runner = self.env.runner
+
+        if environment.web_ui:
+
+            @environment.web_ui.app.route("/csv_results.csv")
+            def csv_results_page():
+                """
+                a different way of obtaining results rather than writing to disk
+                to use it getting all results back, set the flush_size to
+                a high enough value that it will not flush during your test
+                """
+                response = environment.web_ui.app.response_class(
+                    response=self.field_delimiter.join(self.csv_headers)
+                    + self.row_delimiter
+                    + self.row_delimiter.join(self.csv_results),
+                    status=200,
+                    mimetype="text/csv",
+                )
+                return response
+
+    def _create_results_log(self):
+        """
+        creates a results log
+        """
+        results_file = open(self.results_filename, "w")
+        results_file.write(self.field_delimiter.join(self.csv_headers) + self.row_delimiter)
+        results_file.close()
+        return results_file
+
+    def _flush_to_log(self):
+        """
+        flushes results to log file
+        """
+        self.results_file = open(self.results_filename, "a")
+        self.results_file.write(self.row_delimiter.join(self.csv_results) + self.row_delimiter)
+        self.results_file.close()
+        self.csv_results = []
+
+    def _write_final_log(self):
+        """
+        performs final write to log file when test complete
+        """
+        self.results_file = open(self.results_filename, "a")
+        self.results_file.write(self.row_delimiter.join(self.csv_results) + self.row_delimiter)
+        self.results_file.close()
+
+    def start_logging(self, user):
+        self._add_user()
```

Comment:
will this be correct? start_logging is only called once, right? If you want the current number of active users, check runner.user_count

Reply:
Or is this something else (I can see you're using runner.user_count in a different place)

Reply:
The idea behind this for tests with multiple classes. Each will have its own user count but the runner will report the total.

Reply:
Ah, yes that makes sense.